### ISSUE TITLE
RAML Support?

### ISSUE THREAD

**Issue by brylie** [NREL/api-umbrella#47]
What are your thoughts on API Umbrella as it relates to the [RAML specification](http://raml.org/docs.html)?

**Comment by darylrobbins** [comment]
RAML is also fantastic for offering auto generated API documentation. This is how we are planning to document our API.

**Comment by perfaram** [comment]
IMHO, RAML in api-umbrella would be a great addition

**Comment by brylie** [comment]
I set up a [Bountysource page](https://www.bountysource.com/issues/5962299-raml-support) to support this issue financially.

**Comment by kyyberi** [comment]
I would reformat the needs here. RAML style API description is one approach. There are atleast two more significant API description frameworks/specifications: Swagger 2.0 (YAML based as is RAML) and API Blueprint. 

Umbrella should offer means to input specification file in above given formats and set routing and all in Umbrella according to specification. Parsers for RAML, API blueprint and Swagger exists. 

At least API Blueprint has Ruby on Rails binding https://github.com/apiaryio/redsnow as well as RAML https://github.com/coub/raml_ruby

**Comment by darylrobbins** [comment]
I think this issue is currently too nebulous. We should try and be more concrete about what we would expect to happen when we import an API specification file. What would the MVP look like vs. what other features our on the wishlist? I think what an imported API spec means is going to depend a lot on how one is mapping urls to backend api's. Care to offer some concrete examples of what kind of configuration you'd like to be done automatically?

@kyyberi Excellent point, but I think we may want to focus on one spec to start and go from there. Also, unfortunately, the RAML Ruby library is incomplete and not overly useful at the moment.

We have been doing some work on generating HTML documentation from a RAML spec.

**Comment by darylrobbins** [comment]
The ideal scenario would be to only have to support one spec format, which the others could be converted into.

**Comment by kyyberi** [comment]
I agree that one format only support is ideal. Who's going to make the selection which it is? :) I guess RAML is strong candidate. At the moment I don't have a concrete example to offer and currently my time is limited. Therefore I can't promise that I will ever offer it. Anyway starting with minimum read in approach would make sense...and minimize the amount of waste. 

If and when RAML ruby parser is not very useful and if the development seems to lag, perhaps it makes sense to start with some other format. Neverthless, this API-first design driven support will be awesome feature when it's done :)

**Comment by darylrobbins** [comment]
I think whoever ends up implementing the feature gets to decide. :) I agree that RAML is a strong candidate. I see more about it than the other two projects. But yes, it looks like the RAML Ruby parser was left in an incomplete state and hasn't got much love since (it's not an official project). We gave up on the gem pretty early and decided to use JavaScript for most of our RAML needs thereafter.
